### PR TITLE
ChatGPT Auto-Answer should not be posted when image

### LINKED ISSUE #918: ChatGPT Auto-Answer should not be posted when image
ChatGPT should not post an auto-answer attempt when the question contains an image or link.

Since it has no access to reading that, it can only post bad quality answers, which we do not want to post in the first place.

### PR DESCRIPTION
**Pull request by SquidXTV**
# About
resolves #918

Skips generating an auto-answer if message contains images or links.
This is to prevent answers which miss context for ChatGPT to give a decent answer.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -86,4 +86,12 @@ subprojects {
...
         useJUnitPlatform()
     }
+
+    compileJava {
+        options.encoding = "UTF-8"
+    }
+
+    compileTestJava {
+        options.encoding = "UTF-8"
+    }
```
Comment (ankitsmt211) [inline build.gradle:96]: Might be a good idea to make another PR for these changes no? Ideally working on single issue keeps PRs focussed and meaningful.

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Utility class to detect links.
+ */
+public class LinkDetections {
+
+    private LinkDetections() {
```
Comment (ankitsmt211) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: any reason to create a seperate class? or it's just git being dumb when you change name of file?

**Review by ankitsmt211 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java (hunk 2 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: yeah there is a reason, because it is used in multiple places (`LinkPreview` and `HelpThreadCreatedListener`)

File: build.gradle (hunk 1 above)
Comment (SquidXTV) [reply build.gradle:96]: I am not sure about this, I already thought about making a different PR for it but also thought thats just extra work for no benefit actually

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java (hunk 2 above)
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: i mean to say the content seems pretty much same as `LinkPreview`, did you just change name of file or added stuff? since it shows all of the code as new in diff.

File: build.gradle (hunk 1 above)
Comment (ankitsmt211) [reply build.gradle:96]: It's just two lines of code, either way it's out of scope for this PR. Would appreciate if you can address this in new PR.

**Review by ankitsmt211 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java (hunk 2 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: it changed parameter and added one new method

File: build.gradle (hunk 1 above)
Comment (SquidXTV) [reply build.gradle:96]: I don't really see the reason, it's not that important to have a separate PR in my opinion
going to wait what other say about this

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java [hunk 3]
```diff
@@ -2,11 +2,7 @@
 
 import com.github.benmanes.caffeine.cache.Cache;
 import com.github.benmanes.caffeine.cache.Caffeine;
-import net.dv8tion.jda.api.entities.Member;
-import net.dv8tion.jda.api.entities.Message;
-import net.dv8tion.jda.api.entities.MessageEmbed;
-import net.dv8tion.jda.api.entities.Role;
-import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.entities.*;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java:5]: please adjust ur IDE rules. no star imports.

File: build.gradle (hunk 1 above)
Comment (Zabuzard) [reply build.gradle:96]: i get ankis point but i also dont mind too much.
at the end of the day what counts is what ends in master 🤷

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (ankitsmt211) [reply build.gradle:96]: idk it's the end of the world you can very well have all your code in one PR, but ig the idea behind suggestion was to promote more atomic PRs. Whatever we choose do reflect across project in future PRs as well, although no significant impact but just becomes a habbit in general.

Comment (Zabuzard) [reply build.gradle:96]: I just want to avoid people being too annoyed by pure bureaucracy so that the end result is that they wont fix smaller issues as they read them because "damn, im too lazy to put these 3 smaller typos here in a separate PR".

I rather have people fix small shit as they read it than not having the improvements at all. Thats my POV and I have witnessed this issue several times in other (bigger) projects.

But I ofc also agree that in a perfect world, it should be in a separate PR.

Comment (ankitsmt211) [reply build.gradle:96]: very well, if you say  so.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java (hunk 3 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadCreatedListener.java:5]: hm weird, I think this was done by `Optimize Imports` :/

**Review by ankitsmt211 — APPROVED** [review]
looks good to me 👍

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Extracts all links from the given content.
+     *
+     * @param content the content to search through
+     * @param filterSuppressed filters links suppressed with {@literal <url>}
+     * @param filterNonHttpSchemes filters links that are not using http scheme
+     * @return a list of all found links, can be empty
+     */
+    public static List<String> extractLinks(String content, boolean filterSuppressed,
+            boolean filterNonHttpSchemes) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:28]: im putting this as general note, cause its not super important for now, but this is the so called "boolean trap".

to a reader without param-hints, `extractLinks(text, true, false)` is confusing, as they have no idea what the bools mean (without looking up the doc).

a code-way to fix the issue is by introducing quick enums instead. like

```java
enum LinkFilter {
  SUPPRESSED,
  NON_HTTP_SCHEME
}
```
and then instead accept a `Set<LinkFilter> filters` as parameter. the call-site then looks like
```java
extractLinks(text, Set.of(LinkFilter.SUPPRESSED, ...))
```

Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:28]: I knew you would say that.
Thanks for the suggestion, somehow couldn't think of the enum solution myself lol
Would it maybe make more sense using varargs for the parameter?

Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:28]: > im putting this as general note, cause its not super important for now, but this is the so called "boolean trap".
> 
> to a reader without param-hints, `extractLinks(text, true, false)` is confusing, as they have no idea what the bools mean (without looking up the doc).
> 
> a code-way to fix the issue is by introducing quick enums instead. like
> 
> ```java
> enum LinkFilter {
>   SUPPRESSED,
>   NON_HTTP_SCHEME
> }
> ```
> and then instead accept a `Set<LinkFilter> filters` as parameter. the call-site then looks like
> ```java
> extractLinks(text, Set.of(LinkFilter.SUPPRESSED, ...))
> ```

Yea I had some difficulty catching up with flags but then since there was documentation as well, it was okayish for me.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:28]: im not a fan of varargs cause of all the edge cases u would have to handle. In this case:
* null
* duplicates

and `Set.of` is kinda convenient already anyways, fortunately.

but sure, its not a big deal for now

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkFilter.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package org.togetherjava.tjbot.features.utils;
+
+/**
+ * Possible ways to filter a link.
+ * 
+ * @see LinkDetections
+ */
+public enum LinkFilter {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/LinkFilter.java:8]: cool cool. btw, it would have been okay for me if u added it as "nested" to the other class 👍

**Review by Zabuzard — DISMISSED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkFilter.java (hunk 5 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkFilter.java:8]: I wasn't sure lmao
ok going to fix it

**Comment by SquidXTV** [comment]
Why does it "redo" all commits if I just rebase?

**Review by Zabuzard — DISMISSED** [review]

**Comment by Zabuzard** [comment]
@SquidXTV because a rebase rewrites the history (hence it requires a force-push).
i.e. you do not just add something "on top", but you changed something behind (namely your base)

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Utility class to detect links.
+ */
+public class LinkDetections {
+
```
Comment (ankitsmt211) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: shouldn't it named just `LinkDetection`?

File: database/build.gradle [hunk 7]
```diff
@@ -7,7 +7,7 @@ var sqliteVersion = "3.45.0.0"
 dependencies {
     implementation 'com.google.code.findbugs:jsr305:3.0.2'
     implementation "org.xerial:sqlite-jdbc:${sqliteVersion}"
-    implementation 'org.flywaydb:flyway-core:10.6.0'
+    implementation 'org.flywaydb:flyway-core:10.7.1'
```
Comment (ankitsmt211) [inline database/build.gradle:10]: is this because of rebase?

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java (hunk 6 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: I wasnt sure because the other helper class is also called `LinkPreviews`

File: database/build.gradle (hunk 7 above)
Comment (SquidXTV) [reply database/build.gradle:10]: most likely, but it shouldnt show up as a new change right?

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java (hunk 6 above)
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetections.java:15]: imo LinkDetection would be a good fit here. Either way LinkDetections feels a bit odd to read.

File: database/build.gradle (hunk 7 above)
Comment (ankitsmt211) [reply database/build.gradle:10]: maybe if it's based on earlier iteration of main branch, @Zabuzard prolly has some insights here.

Comment (SquidXTV) [reply database/build.gradle:10]: I just checked and the version on `develop` branch is also `10.7.1`, so this "change" is kinda invalid

**Review by ankitsmt211 — APPROVED** [review]
good work 👍

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetection.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * Extracts all links from the given content.
+     *
+     * @param content the content to search through
+     * @param filter the filters applied to the urls
+     * @return a list of all found links, can be empty
+     */
+    public static List<String> extractLinks(String content, Set<LinkFilter> filter) {
+        return new UrlDetector(content, UrlDetectorOptions.BRACKET_MATCH).detect()
```
Comment (Alathreon) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetection.java:44]: Maybe add a line break before detect()

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetection.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Checks whether the given content contains a link.
+     *
+     * @param content the content to search through
+     * @return true if the content contains at least one link
+     */
+    public static boolean containsLink(String content) {
+        return !(new UrlDetector(content, UrlDetectorOptions.BRACKET_MATCH).detect().isEmpty());
```
Comment (Alathreon) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetection.java:58]: The parenthesis is useless here.

File: build.gradle [hunk 10]
```diff
@@ -86,12 +86,4 @@ subprojects {
...
         useJUnitPlatform()
     }
-
-    compileJava {
-        options.encoding = "UTF-8"
-    }
-
-    compileTestJava {
-        options.encoding = "UTF-8"
-    }
```
Comment (Alathreon) [inline build.gradle:96]: It is not normal that you have to fix this, do a rebase.

**Review by Alathreon — DISMISSED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetection.java (hunk 9 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/LinkDetection.java:58]: how is it useless? its needed for `!` I think

File: build.gradle (hunk 10 above)
Comment (SquidXTV) [reply build.gradle:96]: it is, I moved it to a different pr

Comment (SquidXTV) [reply build.gradle:96]: but yeah I need to rebase

**Review by marko-radosavljevic — APPROVED** [review]
Thanks :heart:
